tests/pass-dep/libc/pthread-sync.rs

```diff
@@ -215,6 +214,8 @@ fn check_rwlock_read_no_deadlock() {
     }
 }
 
+// TODO: There is a data race error when there is no -Zmiri-preemption-rate=0, test this again after
+// the ICE is resolved.
 fn check_cond() {
```

Comment:
Surprisingly miri reported data race on this test when there is no ``-Zmiri-preemption-rate=0``. I haven't looked into this yet, but I will revert this change for now as ``-Zmiri-preemption-rate=0`` is needed to trigger the ICE.

Reply:
Before the data race is reported, there is another failed assertion reported, so the data race might be an side effect of the failed assertion? If no one's alarm went off about this, it's probably fine. 
<details>
<summary> error trace </summary>

```rust
thread 'main' panicked at ./tests/pass-dep/libc/pthread-sync.rs:247:9:
assertion failed: data.ptr.read() == 1
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
note: in Miri, you may have to set `MIRIFLAGS=-Zmiri-env-forward=RUST_BACKTRACE` for the environment variable to have an effect
error: Undefined Behavior: Data race detected between (1) atomic store on thread `unnamed-11` and (2) deallocation on thread `main` at alloc23901. (2) just happened here
   --> ./tests/pass-dep/libc/pthread-sync.rs:218:1
    |
218 | / fn check_cond() {
219 | |     unsafe {
220 | |         let mut cond: MaybeUninit<libc::pthread_cond_t> = MaybeUninit::uninit();
221 | |         assert_eq!(libc::pthread_cond_init(cond.as_mut_ptr(), ptr::null()), 0);
...   |
263 | |     }
264 | | }
    | |_^ Data race detected between (1) atomic store on thread `unnamed-11` and (2) deallocation on thread `main` at alloc23901. (2) just happened here
    |
help: and (1) occurred earlier here
   --> ./tests/pass-dep/libc/pthread-sync.rs:234:24
    |
234 |             assert_eq!(libc::pthread_mutex_lock(mutex.ptr), 0); // This is one
    |                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    = help: this indicates a bug in the program: it performed an invalid operation, and caused Undefined Behavior
    = help: see https://doc.rust-lang.org/nightly/reference/behavior-considered-undefined.html for further information
    = note: BACKTRACE (of the first span):
    = note: inside `check_cond` at ./tests/pass-dep/libc/pthread-sync.rs:218:1: 264:2
note: inside `main`
   --> ./tests/pass-dep/libc/pthread-sync.rs:21:5
    |
21  |     check_cond();
    |     ^^^^^^^^^^^^

note: some details are omitted, run with `MIRIFLAGS=-Zmiri-backtrace=full` for a verbose backtrace

error: aborting due to 1 previous error

error: test failed, to rerun pass `--test ui`

```
</details>

But I couldn't understand the diagnostic. ``(2)`` is pointed at the end of the function, which is not really helpful for me to understand where exactly it is. 

```rust
218 | / fn check_cond() {
219 | |     unsafe {
220 | |         let mut cond: MaybeUninit<libc::pthread_cond_t> = MaybeUninit::uninit();
221 | |         assert_eq!(libc::pthread_cond_init(cond.as_mut_ptr(), ptr::null()), 0);
...   |
263 | |     }
264 | | }
    | |_^ Data race detected between (1) atomic store on thread `unnamed-11` and (2) deallocation on thread `main` at alloc23901. (2) just happened here
```

Reply:
What is the status of this? The diff seems to be done, there is no ICE.

> (2) is pointed at the end of the function

Yes, that's where the local variables get deallocated.

Reply:
The reason for the race is that we are panicking, and then we are not executing the `t.join().unwrap()`, and then we get a race.

Reply:
Resolved by https://github.com/rust-lang/miri/pull/4020